### DISCUSSION TITLE
How can I achieve progressive rendering of dynamic routes?

### DISCUSSION THREAD

**Discussion by fabiosantoscode** [react-static/react-static#1537] [Q&A] [answered]
Hello there!

I'm switching away from nextJS because it has too many embedded opinions. However, one thing that I can't get anywhere else, is progressive rendering for dynamic routes.

I noticed that when running `npx react-static create` with the default template, I get a `/dynamic` route, but it's clientside only. When running `yarn build && yarn serve` it's revealed that the 404 page is displayed first, which is to be expected because the path is not returned in `getRoutes`.

However, I need to create some static pages, with URL patterns (eg: `/blog/posts/:postId`), and have them pre-rendered. The clientside code would then pick up the server's HTML and fetch `:postId` from another independent API.

Is it possible for me to create URL patterns in `getRoutes`?

I know I'll probably need a custom server to get the right HTML file based on which pattern matches each route. What I'm not sure about is how I can inform the `<Routes />` component about these patterns, and whether it's even a good idea to attempt to do this.

**Comment by SleeplessByte** [comment, marked as answer]
In order to be able to pre-render and server the posts you either need to know the post ids before, or add a _server_ in front of the serving. So what you could do is something like: have `/blog/posts/__prerender` exist (so it will generate `/blog/posts/__prerender.html` in the output, and serve _that_ file when `/blog/posts/:postid` is requested. 

I don't think react-static provided another way to accomplish this; it's really not the task of `react-static`.

**Reply by fabiosantoscode** [reply]
Adding this __prerender is something akin to what I'm trying to do atm. It's been troublesome though, because I need to tell react-static that `blog/posts/*` is actually supposed to be filled in by this `__prerender`.

However since you point out that it's not `react-static`'s task I'll keep looking for some package which can do it. The reason I'm ditching NextJS is that I'm trying to do things NextJS is not supposed to do :) 

Thanks a lot for this answer!

**Reply by SleeplessByte** [reply]
This might sound weird, but have you considered non-SPA react? So in this case it could be a static HTML file that has all the building blocks to render the page (pre-rendered, as it's HTML), and then a server, such as express, or something rack-enabled (sinatra/rails), or even apache/nginx? All of them could route `/blog/posts/*` to that file. You'd then pull out the last part of the path and do whatever you need to do (grab the post, and render it).

It sounds like that might just be easier than any/most packages you'll find.

**Reply by fabiosantoscode** [reply]
Yes, non-SPA is definitely the space I'm in.

I want nice code splitting, because some of our pages are really heavy in terms of code, and I also want a separate pre-render per page type.

I've been accomplishing this with nextjs by using their filesystem router but not using their data fetching facilities (I want a dumb server).
